libraries/lib-wave-track/CachingPlayableSequence.cpp

```diff
@@ -0,0 +1,115 @@
+#include "CachingPlayableSequence.h"
+#include "AudioSegmentSampleView.h"
+#include "WaveTrack.h"
+
+#include <cassert>
+
+namespace
+{
+void FillBufferFromTrackBlockSequence(
+   const TrackSampleView& track, float* buffer, size_t bufferSize)
+{
+   size_t written = 0;
+   for (const auto& segment : track)
+   {
+      size_t toWrite = std::min(
+         segment.GetSampleCount().as_long_long(),
+         static_cast<long long>(bufferSize - written));
+      written += segment.Copy(buffer, toWrite);
+   }
+   std::fill(buffer + written, buffer + bufferSize, 0.f);
+}
+} // namespace
+
+CachingPlayableSequence::CachingPlayableSequence(const WaveTrack& waveTrack)
+    : mWaveTrack(waveTrack)
+{
+}
+
+size_t CachingPlayableSequence::NChannels() const
+{
+   return mWaveTrack.NChannels();
+}
+
+float CachingPlayableSequence::GetChannelGain(int channel) const
+{
+   return mWaveTrack.GetChannelGain(channel);
+}
+
+size_t CachingPlayableSequence::GetMaxBlockSize() const
+{
+   return mWaveTrack.GetMaxBlockSize();
+}
+
+sampleCount CachingPlayableSequence::GetBlockStart(sampleCount t) const
+{
+   return mWaveTrack.GetBlockStart(t);
+}
+
+size_t CachingPlayableSequence::GetBestBlockSize(sampleCount t) const
+{
+   return mWaveTrack.GetBestBlockSize(t);
+}
+
+bool CachingPlayableSequence::Get(
+   size_t iChannel, size_t nBuffers, samplePtr buffers[], sampleFormat format,
+   sampleCount start, size_t len, fillFormat fill, bool mayThrow,
+   sampleCount* pNumWithinClips) const
+{
+   mCacheHolders = mWaveTrack.GetSampleView(iChannel, nBuffers, start, len);
```

Comment:
You assume after this a post of `GetSampleView`:  `size() == nBuffers - iChannel`

Please make that explicit in header comments, and assert the post here

Reply:
`nBuffers - iChannel`, really?
I have been interpreting those arguments as "give me `nBuffers` starting from index `iChannel`", which results in `size() == nBuffers`. What you are saying is "give me buffers from index `iChannel` to index `nBuffers` (exclusive)".
(Your implementation of `WaveTrack::Get` does neither, but I think it attempted the former interpretation. I'm attempting a fix accordingly.)

Reply:
Sorry, what I wrote was mistaken.

What I was thinking was, how to prove you don't index `buffers` out of bounds.

So what I should have said is, there is an unstated postcondition that `GetSampleView` always returns a vector of size `nBuffers`.

So it would be good to make that explicit in that function's header.

Reply:
Ok.
I already added a `@return` line to that header stating this. Is that good enough?